x500/cert/acme/src/main/java/org/wildfly/security/certificate/management/x500/cert/acme/AsyncAcmeClientSpi.java

```diff
@@ -0,0 +1,282 @@
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.time.Instant;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+
+import static org.wildfly.security.certificate.management.x500.cert.acme.Acme.ACCEPT_LANGUAGE;
+import static org.wildfly.security.certificate.management.x500.cert.acme.Acme.BAD_NONCE;
+import static org.wildfly.security.certificate.management.x500.cert.acme.Acme.BASE64_URL;
+import static org.wildfly.security.certificate.management.x500.cert.acme.Acme.INSTANCE;
+import static org.wildfly.security.certificate.management.x500.cert.acme.Acme.JSON_CONTENT_TYPE;
+import static org.wildfly.security.certificate.management.x500.cert.acme.Acme.PROBLEM_JSON_CONTENT_TYPE;
+import static org.wildfly.security.certificate.management.x500.cert.acme.Acme.RATE_LIMITED;
+import static org.wildfly.security.certificate.management.x500.cert.acme.Acme.REPLAY_NONCE;
+import static org.wildfly.security.certificate.management.x500.cert.acme.Acme.RETRY_AFTER;
+import static org.wildfly.security.certificate.management.x500.cert.acme.Acme.TYPE;
+import static org.wildfly.security.certificate.management.x500.cert.acme.Acme.USER_ACTION_REQUIRED;
+import static org.wildfly.security.certificate.management.x500.cert.acme.Acme.USER_AGENT;
+import static org.wildfly.security.certificate.management.x500.cert.acme.CertMgmtMessages.acme;
+
+public abstract class AsyncAcmeClientSpi {
+
+    private static final long DEFAULT_RETRY_AFTER_MILLI = 3000;
+    private static final String USER_AGENT_STRING = "Elytron ACME Client/" + "1.0.0.Alpha1-SNAPSHOT";
+    private final AcmeClientSpi delegate;
+
+    public AsyncAcmeClientSpi() {
+        AsyncAcmeClientSpi impl = this;
+
+        this.delegate = new AcmeClientSpi() {
+            @Override
+            public AcmeChallenge proveIdentifierControl(AcmeAccount account, List<AcmeChallenge> challenges) throws AcmeException {
+                return impl.proveIdentifierControl(account, challenges);
+            }
+
+            @Override
+            public void cleanupAfterChallenge(AcmeAccount account, AcmeChallenge challenge) throws AcmeException {
+                impl.cleanupAfterChallenge(account, challenge);
+            }
+        };
+    }
+
+    protected abstract AcmeChallenge proveIdentifierControl(AcmeAccount account, List<AcmeChallenge> challenges) throws AcmeException;
+
+    protected abstract void cleanupAfterChallenge(AcmeAccount account, AcmeChallenge challenge) throws AcmeException;
+
+    public Uni<byte[]> getNewNonce(final AcmeAccount account, final boolean staging) {
+        if (account == null) {
+            return Uni.createFrom().failure(new IllegalArgumentException("account"));
+        }
+        HttpClient httpClient = HttpClient.newHttpClient();
```

Comment:
Unrelated to this PR but Quarkus is avoiding using the JDK client.
Should we consider adding an SPI?

Reply:
Quarkus Keycloak Authorization uses Keycloak authorization HttpRequest/HttpResponse SPI, see https://github.com/quarkusio/quarkus/blob/main/extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/runtime/VertxHttpFacade.java, may be something similar can be considered

Reply:
@cescoffier @sberyozkin  Thank you, I have created HttpRequestSpi, HttpResponseSpi, and HttpClientSpi so that any other client could be used. You can see how they look [here](https://github.com/Skyllarr/certificate-management/tree/async-spi-with-rewrite/x500/cert/spi/src/main/java/org/wildfly/security/certificate/management/x500/cert/spi).

In my tests I used standard JDK client for now. Which HTTP client would you prefer? I can use the preferred one in my tests instead

Reply:
We will use the Vert.x Web client exposing the Mutiny API: https://github.com/smallrye/smallrye-mutiny-vertx-bindings/blob/main/vertx-mutiny-clients/vertx-mutiny-web-client/src/test/java/io/vertx/mutiny/web/WebClientTest.java#L41